### DISCUSSION TITLE
Can't coalesce video tracks like the manual shows

### DISCUSSION THREAD

**Discussion by Evan8080** [savonet/liquidsoap#3511] [User Support]
The instructions to [add a default video track](/savonet/liquidsoap/blob/main/doc/content/multitrack.md#add-a-default-video-track) don't work as expected.

With the following code:

```liquidsoap
cover = single('image.png')
video = single('video.mkv')

ctrack = source.tracks(cover)
vtrack = source.tracks(video)

result = source({
	audio    = vtrack.audio,
	video    = vtrack.video ?? ctrack.video,
	metadata = vtrack.metadata
})

encode  = %ffmpeg(format="mpegts", %video(codec="libx264"), %audio(codec="aac", channels=2, ar=44100))
streams = [("default", encode)]

output.file.hls("/srv/http", streams, result)
```

The video track always ends up being `ctrack`'s video despite `vtrack` having video already.

I have tried the exact form from the manual and over a dozen other tests before reporting, all with the same result.

Using the official `docker.io/savonet/liquidsoap-alpine:v2.2.2`

**Comment by Evan8080** [comment]
@toots user support? the exact same example from the manual won't work at all, in fact I spent a whole day searching online and in the book for hints and could not find any.

Is there really anything there that he user should understand that I don't? The coalescing operator works correctly with other values as the manual example suggest and as it makes sense with the operator.

This is in fact the only thing I really want to do with Liquidsoap, everything else works except this example in particular.

**Reply by toots** [reply]
I'm looking at it.

**Reply by toots** [reply]
Yeah it's a little embarrassing to admin but I'm not sure what happened around the time I wrote this doc and when the final feature set landed. The feature does not work at the moment. I have ideas to make it work, though and will see about a quick follow-up.

If you give

**Reply by Evan8080** [reply]
>  If you give

If I give...? 🤔

I just tried again with the latest version, following the documentation:

```cover = single('image.png')
list  = mksafe(playlist('any'))

video  = source.tracks(list).video ?? source.tracks(cover).video
result = source(source.tracks(list).{video=video})

encode  = %ffmpeg(format="mpegts", %video(codec="libx264"), %audio(codec="aac", channels=2, ar=44100))
streams = [("default", encode)]

output.file.hls("/srv/http", streams, result)
```

`list` will contain a single video file, which otherwise works alright, but here the cover image is always used instead of the single video in the playlist.

Expected result would be to see the video of the currently playing file and only use the cover as fallback video if no video is available.

When starting Liquidsoap says that `cover` isn't used despite being actually used as shown by the test results.

```
At main.liq, line 1, char 0-27:                                                                                                            
cover = single('image.png')

Warning 4: Unused variable cover
```

using `ghcr.io/savonet/liquidsoap-alpine:c3f311e-minimal`

**Reply by vitoyucepi** [reply]
Hi @Evan8080,
I think something is wrong with the typing system.

```ruby
ctrack = source.tracks(cover)
vtrack = source.tracks(video)

result = source({
  audio = vtrack.audio,
  video = vtrack.video ?? ctrack.video,
  metadata = vtrack.metadata
})
```
Liquidsoap will resolve `video` without video.
```
Requested content-type for "video.mkv": {audio=pcm(stereo)}
```

You should set the format explicitly. But I don't know if this solution works perfectly in all cases.
```ruby
cover = (single("image.png") : source(video=canvas))
video = (single("video.mkv") : source(audio=pcm(stereo), video=canvas))
```

**Reply by toots** [reply]
t looks like you are trying to say: "play the video track from `vtrack` if it exists, `ctrack` otherwise. This is a feature that is not available at the moment, all sources must have a fixed expected content. I would love to add it at some point, though.

**Reply by evaninf** [reply]
Is there any way to do this at all?

I've been trying with dynamic sources,  requests, protocols... anything I find; nothing seems to work and most of the time error messages are too cryptic to make any sense of them.

Is it even possible to tell when a source lacks video tracks? I know... it should be possible but I just can't find a way, there are no examples and the documentation suggests it's possible but even its only example won't work.

**Comment by Evan8080** [comment]
> This is a feature that is not available at the moment, all sources must have a fixed expected content. I would love to add it at some point, though.

This can still be achieved with a custom protocol, right?

**Reply by evaninf** [reply]
> This can still be achieved with a custom protocol, right?

...@toots?

**Reply by toots** [reply]
Sure. You could have a protocol taking the file and adding a default video track using ffmpeg typically.
